### ISSUE TITLE
With 1,000 ssh connections, 426 lost after five days

### ISSUE THREAD

**Issue by neo954** [xcat2/goconserver#9]
This bug is against `goconserver` `Version: 0.1, BuildTime: 2017-11-07T23:00:03-0500 Commit: 906cd810f009dd1b0b22a75ea0c4efc1cbb2e840` running on a ppc64le Red Hat 7.4 Linux environment.

**The recreation steps**

```
# for node in foo{000..999}
> do
>     congo create ${node} driver=ssh ondemand=false \
>         --params user=foo,host=10.3.1.9,port=22,password=something
>     ln -s /dev/null /var/log/goconserver/nodes/${node}.log
>     congo logging ${node} on
> done
```

I created 1,000 node definitions in `congo`. On the ssh server side, a bash script will run after the user log in, which will generate outputs continuously. Plus, the console log files, `/var/log/goconserver/nodes/foo{000..999}.log` are all symbolic linked to `/dev/null`. In order to avoid fill up the `/var` file system.

```
# ps auwx
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
<<<...omit...>>>
root     18953 42.3  0.4 514368 93696 ?        Ssl  05:02   0:07 /usr/local/bin/goconserver
<<<...omit...>>>
```
There are 1,000 active ssh connections.
```
# lsof -p 18953 | grep TCP | grep ssh | grep ESTABLISHED | wc -l
1000
```

Then, I left this test environment run for a handful of days.

**What actually happened**

After around five days...

See the `TIME` slot of the `ps` output. It shows `goconserver` run for about 1,532 minutes and 41 seconds "accumulated cpu time".

```
# ps auwx
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
<<<...omit...>>>
root     18953 20.1  0.6 781120 129664 ?       Ssl  Nov08 1532:41 /usr/local/bin/goconserver
<<<...omit...>>>
```
I noticed some of the ssh connections were disconnected.
```
# lsof -p 18953 | grep TCP | grep ssh | grep ESTABLISHED | wc -l
574
```

**What is expected**

* The common used OpenSSH ssh client has a client side configuration `TCPKeepAlive`, which can be set to `yes`. I quote the relevant part of `man ssh_config` below. Maybe, similar thing can be implemented in `goconserver`.

```
     TCPKeepAlive
             Specifies whether the system should send TCP keepalive messages
             to the other side.  If they are sent, death of the connection or
             crash of one of the machines will be properly noticed.  However,
             this means that connections will die if the route is down tempo‐
             rarily, and some people find it annoying.

             The default is yes (to send TCP keepalive messages), and the
             client will notice if the network goes down or the remote host
             dies.  This is important in scripts, and many users want it too.

             To disable TCP keepalive messages, the value should be set to no.
```

**Comment by chenglch** [comment]
@neo954, could you paste some warning or error level log from `goconserver` to help locate this issue?

**Comment by chenglch** [comment]
@neo954 , sorry for the error append with git,  the latest release has been uploaded. The fix for this issue has been included in 0588fc3c2a84a718634f7d7ad48536e97ec252bf, you could try it again if you have time.
